README.md
# HumanTime
Human Time is a function to convert a string such as "every other tuesday", "every weekday" or "every friday at 2pm" and convert it into a sequence of date times as allowed by the string. Adapted from my [human_time library for Python](https://github.com/Teifion/human_time_py).

## Usage
Human Time can parse both repeating intervals (e.g. "every other day") and also one off relative times (e.g. "next tuesday at 3pm").

```elixir
  HumanTime.repeating("Every wednesday at 1530")
  |> Stream.take(3)
  |> Enum.to_list

  > [
      #DateTime<2018-08-15 15:30:00Z>,
      #DateTime<2018-08-22 15:30:00Z>,
      #DateTime<2018-08-29 15:30:00Z>
    ]

  HumanTime.relative("Next wednesday at 1530")

  > {:ok, #DateTime<2018-08-15 15:30:00.848218Z>}
```

If a time is not part of the string, datetimes will be outputted as being at midnight. You can set a time such as "at 1500" or "at noon" but you can also say "at the current time" or "at this time" and it will take the time of parsing and repeat that.

### Example formats - Repeating
every 5 seconds/minutes/hours/days/weeks/months/years
every weekday at midnight
every weekend at 1500
first monday of every month
second wednesday of every month at midnight
last Friday of every month at 9am
15th of every month at midnight
every other Sunday at this time
first monday after second sunday of every month at midnight

### Example formats - Relative
5m
15/04/2020
2020-04-15 10:15:00
next wednesday at 1500
this time next friday
week next friday at 2am


For a more complete list of examples, check out the [relative](test/relative_test.exs) and [repeating](test/repeating_test.exs) test files.

### Options
`from`: Dictates the date the sequence will begin from (defaults to starting from now)  
`until`: Dictates the date the sequence will end at (defaults to never finish)

## Installation
To add Human Time to your project, you only need to add it as a dependency and start it as an application in your mix.exs file.

```elixir
def deps do
  [
    {:human_time, "~> 0.2.3"}
  ]
end

defp application do
  [applications: [:human_time]]
end
```

## Roadmap / TODO
- Add a debug mode
- Allow for combining strings with "and", e.g. "every friday and every other thursday"
- Allow for exceptions with "except", e.g. "every weekday except fridays"
- Optimise the function calls, some of the filters will iterate through every day in a month for every day in a month
- Add time jumps such as "25th minute of every other hour"


## License

This software is licensed under [the MIT license](LICENSE.md).

CHANGELOG.md
# Change Log

## 0.3.1
- Added support for X months and X years

## 0.3.0
- Added support for "never"
- Used mix.format

## 0.2.4
- Added support for every Xth (e.g. every third) as a skip function
- Added test for "Every other hour" but currently fails as I'm not sure what the correct outcome should be
- Added ability to have spaces between the time and 12hour cycle, e.g. "12 pm"

## 0.2.3
- Added following handlers for repeating: X seconds
- Added following handlers for relative: "now", "In a hour", "In an hour"

## 0.2.2
- Fixed bug where time before date would get seen as just the date

## 0.2.1
- Added repeating option for N of unit X, e.g. "5 months"

## 0.2.0
- Existing functionality is now called "repeadetly"
- Added new set of functionality called "relative"

## 0.1.0
- Released on Hex

lib/human_time.ex
defmodule HumanTime do
  @moduledoc """
  Human Time is a function to convert a string such as "every other tuesday", "every weekday" or "next friday at 2pm" and convert it into a one or a sequence of date times as allowed by the string.
  """

  # TODO: Add in cron-tab formatting (for repeating stuff)
  # Some specific examples that demonstrate the full range of expressions:

  #   0 * * * * — The first minute of every hour
  #   */15 9-17 * * * — Every fifteen minutes during standard business hours
  #   0 0 * DEC * — Once a day at midnight during december
  #   0 7-9,4-6 13 * FRI — Once an hour during both rush hours on Friday the 13th

  alias HumanTime.Repeating
  alias HumanTime.Relative
  require Logger

  @doc """
  Generates a stream of datetimes for the string given.

  ## Options
  `from` The datetime from when the sequence will be generated, defaults to the current time.

  `until` The datetime when the sequence will be terminated, defaults to nil. When nil the sequence will never be terminated.

  ## Example
      HumanTime.repeating("Every wednesday at 1530")
      |> Stream.take(3)
      |> Enum.to_list

      #=> [
      #=>   #DateTime<2018-08-15 15:30:00.848218Z>,
      #=>   #DateTime<2018-08-22 15:30:00.848218Z>,
      #=>   #DateTime<2018-08-29 15:30:00.848218Z>
      #=> ]

  ## Error result is:
    {:error, message}
  """
  @spec repeating(String.t(), [term]) :: {:ok, Enumerable.t()} | {:error, String.t()}
  def repeating(timestring, opts \\ []) do
    # Logger.debug "Calling repeating with timestring: #{timestring} and opts: #{Kernel.inspect opts}"

    from = opts[:from] || Timex.now()
    until = opts[:until]

    while_function = Repeating.Generators.while_function(until)
    result = Repeating.Parser.build_functions(timestring)

    case result do
      {:error, msg} ->
        {:error, msg}

      {:ok, {generator_function, filter_function, mapper_function}} ->
        {:ok,
         from
         |> Stream.iterate(generator_function)
         |> Stream.take_while(while_function)
         |> Stream.filter(filter_function)
         |> Stream.map(mapper_function)
         |> Stream.filter(&(&1 != nil))}
    end
  end

  @doc """
  Repeats the time string and raises an exception in case of errors
  """
  @spec repeating!(String.t(), [term]) :: Enumerable.t()
  def repeating!(timestring, opts \\ []) do
    case repeating(timestring, opts) do
      {:ok, enumerable} -> enumerable
      {:error, msg} -> raise msg
    end
  end

  @doc """
  Generates a single datetime for the string given.

  ## Options
  `from` The datetime from when the sequence will be generated, defaults to the current time.

  ## Example
      HumanTime.relative("Next wednesday at 1530")

      #=> {:ok, #DateTime<2018-08-15 15:30:00.848218Z>}

  ## Error result is:
      #=> {:error, message}
  """
  @spec relative(String.t(), [term]) :: {:ok, DateTime.t() | nil} | {:error, String.t()}
  def relative(timestring, opts \\ []) do
    from = opts[:from] || Timex.now()

    Relative.Parser.parse(timestring, from)
  end

  @doc """
  Creates the time string and raises an exception in case of errors
  """
  @spec relative!(String.t(), [term]) :: DateTime.t() | nil
  def relative!(timestring, opts \\ []) do
    case relative(timestring, opts) do
      {:ok, the_datetime} -> the_datetime
      {:error, msg} -> raise msg
    end
  end
end
